# Восстановление золота из руды

Подготовьте прототип модели машинного обучения для «Цифры». Компания разрабатывает решения для эффективной работы промышленных предприятий.

Модель должна предсказать коэффициент восстановления золота из золотосодержащей руды. Используйте данные с параметрами добычи и очистки. 

Модель поможет оптимизировать производство, чтобы не запускать предприятие с убыточными характеристиками.

Вам нужно:

1. Подготовить данные;
2. Провести исследовательский анализ данных;
3. Построить и обучить модель.

Чтобы выполнить проект, обращайтесь к библиотекам *pandas*, *matplotlib* и *sklearn.* Вам поможет их документация.

## Подготовка данных

In [1]:
#импортируем всё, что может понадобиться
import warnings
warnings.filterwarnings('ignore')

import pandas as pd 

import seaborn as sns

import matplotlib.pyplot as plt

import numpy as np

import plotly.express as px
import plotly.graph_objects as go
import plotly.figure_factory as ff

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import GridSearchCV 
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.dummy import DummyRegressor
from sklearn.metrics import make_scorer
from sklearn.model_selection import cross_val_score
import plotly.figure_factory as ff

In [2]:
#загрузим данные на любой платформе
try:
    df_train = pd.read_csv('gold_recovery_train_new.csv')
    df_test = pd.read_csv('gold_recovery_test_new.csv')
    df_full = pd.read_csv('gold_recovery_full_new.csv')
except:
    df_train = pd.read_csv('/datasets/gold_recovery_train_new.csv')
    df_test = pd.read_csv('/datasets/gold_recovery_test_new.csv')
    df_full = pd.read_csv('/datasets/gold_recovery_full_new.csv')
display(df_full.head())
display(df_train.head())
display(df_test.head())

date  final.output.concentrate_ag  \
0  2016-01-15 00:00:00                     6.055403   
1  2016-01-15 01:00:00                     6.029369   
2  2016-01-15 02:00:00                     6.055926   
3  2016-01-15 03:00:00                     6.047977   
4  2016-01-15 04:00:00                     6.148599   

   final.output.concentrate_pb  final.output.concentrate_sol  \
0                     9.889648                      5.507324   
1                     9.968944                      5.257781   
2                    10.213995                      5.383759   
3                     9.977019                      4.858634   
4                    10.142511                      4.939416   

   final.output.concentrate_au  final.output.recovery  final.output.tail_ag  \
0                    42.192020              70.541216             10.411962   
1                    42.701629              69.266198             10.462676   
2                    42.657501              68.116445             10.507046   
3                    42.689819              68.347543             10.422762   
4                    42.774141              66.927016             10.360302   

   final.output.tail_pb  final.output.tail_sol  final.output.tail_au  ...  \
0              0.895447              16.904297              2.143149  ...   
1              0.927452              16.634514              2.224930  ...   
2              0.953716              16.208849              2.257889  ...   
3              0.883763              16.532835              2.146849  ...   
4              0.792826              16.525686              2.055292  ...   

   secondary_cleaner.state.floatbank4_a_air  \
0                                 14.016835   
1                                 13.992281   
2                                 14.015015   
3                                 14.036510   
4                                 14.027298   

   secondary_cleaner.state.floatbank4_a_level  \
0                                 -502.488007   
1                                 -505.503262   
2                                 -502.520901   
3                                 -500.857308   
4                                 -499.838632   

   secondary_cleaner.state.floatbank4_b_air  \
0                                 12.099931   
1                                 11.950531   
2                                 11.912783   
3                                 11.999550   
4                                 11.953070   

   secondary_cleaner.state.floatbank4_b_level  \
0                                 -504.715942   
1                                 -501.331529   
2                                 -501.133383   
3                                 -501.193686   
4                                 -501.053894   

   secondary_cleaner.state.floatbank5_a_air  \
0                                  9.925633   
1                                 10.039245   
2                                 10.070913   
3                                  9.970366   
4                                  9.925709   

   secondary_cleaner.state.floatbank5_a_level  \
0                                 -498.310211   
1                                 -500.169983   
2                                 -500.129135   
3                                 -499.201640   
4                                 -501.686727   

   secondary_cleaner.state.floatbank5_b_air  \
0                                  8.079666   
1                                  7.984757   
2                                  8.013877   
3                                  7.977324   
4                                  7.894242   

   secondary_cleaner.state.floatbank5_b_level  \
0                                 -500.470978   
1                                 -500.582168   
2                                 -500.517572   
3                                 -500.255908   
4                                 -500.356035   

   secondary_cleaner.state.floatbank6_a_air  \
0    

date  final.output.concentrate_ag  \
0  2016-01-15 00:00:00                     6.055403   
1  2016-01-15 01:00:00                     6.029369   
2  2016-01-15 02:00:00                     6.055926   
3  2016-01-15 03:00:00                     6.047977   
4  2016-01-15 04:00:00                     6.148599   

   final.output.concentrate_pb  final.output.concentrate_sol  \
0                     9.889648                      5.507324   
1                     9.968944                      5.257781   
2                    10.213995                      5.383759   
3                     9.977019                      4.858634   
4                    10.142511                      4.939416   

   final.output.concentrate_au  final.output.recovery  final.output.tail_ag  \
0                    42.192020              70.541216             10.411962   
1                    42.701629              69.266198             10.462676   
2                    42.657501              68.116445             10.507046   
3                    42.689819              68.347543             10.422762   
4                    42.774141              66.927016             10.360302   

   final.output.tail_pb  final.output.tail_sol  final.output.tail_au  ...  \
0              0.895447              16.904297              2.143149  ...   
1              0.927452              16.634514              2.224930  ...   
2              0.953716              16.208849              2.257889  ...   
3              0.883763              16.532835              2.146849  ...   
4              0.792826              16.525686              2.055292  ...   

   secondary_cleaner.state.floatbank4_a_air  \
0                                 14.016835   
1                                 13.992281   
2                                 14.015015   
3                                 14.036510   
4                                 14.027298   

   secondary_cleaner.state.floatbank4_a_level  \
0                                 -502.488007   
1                                 -505.503262   
2                                 -502.520901   
3                                 -500.857308   
4                                 -499.838632   

   secondary_cleaner.state.floatbank4_b_air  \
0                                 12.099931   
1                                 11.950531   
2                                 11.912783   
3                                 11.999550   
4                                 11.953070   

   secondary_cleaner.state.floatbank4_b_level  \
0                                 -504.715942   
1                                 -501.331529   
2                                 -501.133383   
3                                 -501.193686   
4                                 -501.053894   

   secondary_cleaner.state.floatbank5_a_air  \
0                                  9.925633   
1                                 10.039245   
2                                 10.070913   
3                                  9.970366   
4                                  9.925709   

   secondary_cleaner.state.floatbank5_a_level  \
0                                 -498.310211   
1                                 -500.169983   
2                                 -500.129135   
3                                 -499.201640   
4                                 -501.686727   

   secondary_cleaner.state.floatbank5_b_air  \
0                                  8.079666   
1                                  7.984757   
2                                  8.013877   
3                                  7.977324   
4                                  7.894242   

   secondary_cleaner.state.floatbank5_b_level  \
0                                 -500.470978   
1                                 -500.582168   
2                                 -500.517572   
3                                 -500.255908   
4                                 -500.356035   

   secondary_cleaner.state.floatbank6_a_air  \
0    

date  primary_cleaner.input.sulfate  \
0  2016-09-01 00:59:59                     210.800909   
1  2016-09-01 01:59:59                     215.392455   
2  2016-09-01 02:59:59                     215.259946   
3  2016-09-01 03:59:59                     215.336236   
4  2016-09-01 04:59:59                     199.099327   

   primary_cleaner.input.depressant  primary_cleaner.input.feed_size  \
0                         14.993118                         8.080000   
1                         14.987471                         8.080000   
2                         12.884934                         7.786667   
3                         12.006805                         7.640000   
4                         10.682530                         7.530000   

   primary_cleaner.input.xanthate  primary_cleaner.state.floatbank8_a_air  \
0                        1.005021                             1398.981301   
1                        0.990469                             1398.777912   
2                        0.996043                             1398.493666   
3                        0.863514                             1399.618111   
4                        0.805575                             1401.268123   

   primary_cleaner.state.floatbank8_a_level  \
0                               -500.225577   
1                               -500.057435   
2                               -500.868360   
3                               -498.863574   
4                               -500.808305   

   primary_cleaner.state.floatbank8_b_air  \
0                             1399.144926   
1                             1398.055362   
2                             1398.860436   
3                             1397.440120   
4                             1398.128818   

   primary_cleaner.state.floatbank8_b_level  \
0                               -499.919735   
1                               -499.778182   
2                               -499.764529   
3                               -499.211024   
4                               -499.504543   

   primary_cleaner.state.floatbank8_c_air  ...  \
0                             1400.102998  ...   
1                             1396.151033  ...   
2                             1398.075709  ...   
3                             1400.129303  ...   
4                             1402.172226  ...   

   secondary_cleaner.state.floatbank4_a_air  \
0                                 12.023554   
1                                 12.058140   
2                                 11.962366   
3                                 12.033091   
4                                 12.025367   

   secondary_cleaner.state.floatbank4_a_level  \
0                                 -497.795834   
1                                 -498.695773   
2                                 -498.767484   
3                                 -498.350935   
4                                 -500.786497   

   secondary_cleaner.state.floatbank4_b_air  \
0                                  8.016656   
1                                  8.130979   
2                                  8.096893   
3                                  8.074946   
4                                  8.054678   

   secondary_cleaner.state.floatbank4_b_level  \
0                                 -501.289139   
1                                 -499.634209   
2                                 -500.827423   
3                                 -499.474407   
4                                 -500.397500   

   secondary_cleaner.state.floatbank5_a_air  \
0                                  7.946562   
1                                  7.958270   
2                                  8.071056   
3                                  7.897085   
4                                  8.107890   

   secondary_cleaner.state.floatbank5_a_level  \
0                                 -432.317850   
1                                 -525.839648   
2                                 -500.801673   
3     

In [3]:
# посмотрим на них. Будет TLDR, но зато всё одним разом
pd.set_option('display.max_rows', None) #датафрейм большой, сделаем, чтобы смотрелись все столбцы
names= [df_train, df_test, df_full]
names2= ['df_train', 'df_test', 'df_full']
def info_func(names):
    for name in names:
        df_temp = pd.DataFrame(name)
        display(df_temp.head(10))
        display(df_temp.info())
        display('Кол-во дублей:', df_temp.duplicated().sum())
        display('Кол-во НАНов:', df_temp.isna().sum())
        display('Статистика:', df_temp.describe(),
              '=======================================================================================================')
    return
info_func(names)

date  final.output.concentrate_ag  \
0  2016-01-15 00:00:00                     6.055403   
1  2016-01-15 01:00:00                     6.029369   
2  2016-01-15 02:00:00                     6.055926   
3  2016-01-15 03:00:00                     6.047977   
4  2016-01-15 04:00:00                     6.148599   
5  2016-01-15 05:00:00                     6.482968   
6  2016-01-15 06:00:00                     6.533849   
7  2016-01-15 07:00:00                     6.130823   
8  2016-01-15 08:00:00                     5.834140   
9  2016-01-15 09:00:00                     5.687063   

   final.output.concentrate_pb  final.output.concentrate_sol  \
0                     9.889648                      5.507324   
1                     9.968944                      5.257781   
2                    10.213995                      5.383759   
3                     9.977019                      4.858634   
4                    10.142511                      4.939416   
5                    10.049416                      5.480257   
6                    10.058141                      4.569100   
7                     9.935481                      4.389813   
8                    10.071156                      4.876389   
9                     9.980404                      5.282514   

   final.output.concentrate_au  final.output.recovery  final.output.tail_ag  \
0                    42.192020              70.541216             10.411962   
1                    42.701629              69.266198             10.462676   
2                    42.657501              68.116445             10.507046   
3                    42.689819              68.347543             10.422762   
4                    42.774141              66.927016             10.360302   
5                    41.633678              69.465816             10.182708   
6                    41.995316              69.300835             10.304598   
7                    42.452727              70.230976             10.443288   
8                    43.404078              69.688595             10.420140   
9                    43.235220              70.279619             10.487013   

   final.output.tail_pb  final.output.tail_sol  final.output.tail_au  ...  \
0              0.895447              16.904297              2.143149  ...   
1              0.927452              16.634514              2.224930  ...   
2              0.953716              16.208849              2.257889  ...   
3              0.883763              16.532835              2.146849  ...   
4              0.792826              16.525686              2.055292  ...   
5              0.664118              16.999638              1.918586  ...   
6              0.807342              16.723575              2.058913  ...   
7              0.949346              16.689959              2.143437  ...   
8              1.065453              17.201948              2.209881  ...   
9              1.159805              17.483979              2.209593  ...   

   secondary_cleaner.state.floatbank4_a_air  \
0                                 14.016835   
1                                 13.992281   
2                                 14.015015   
3                                 14.036510   
4                                 14.027298   
5                                 13.938497   
6                                 14.046819   
7                                 13.974691   
8                                 13.964030   
9                                 13.989632   

   secondary_cleaner.state.floatbank4_a_level  \
0                                 -502.488007   
1                                 -505.503262   
2                                 -502.520901   
3                                 -500.857308   
4                                 -499.838632   
5                                 -500.970168   
6                                 -500.971133   
7                                 -501.819696   
8                             

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14149 entries, 0 to 14148
Data columns (total 87 columns):
 #   Column                                              Non-Null Count  Dtype  
---  ------                                              --------------  -----  
 0   date                                                14149 non-null  object 
 1   final.output.concentrate_ag                         14148 non-null  float64
 2   final.output.concentrate_pb                         14148 non-null  float64
 3   final.output.concentrate_sol                        13938 non-null  float64
 4   final.output.concentrate_au                         14149 non-null  float64
 5   final.output.recovery                               14149 non-null  float64
 6   final.output.tail_ag                                14149 non-null  float64
 7   final.output.tail_pb                                14049 non-null  float64
 8   final.output.tail_sol                               14144 non-null  float64


None

'Кол-во дублей:'

0

'Кол-во НАНов:'

date                                                     0
final.output.concentrate_ag                              1
final.output.concentrate_pb                              1
final.output.concentrate_sol                           211
final.output.concentrate_au                              0
final.output.recovery                                    0
final.output.tail_ag                                     0
final.output.tail_pb                                   100
final.output.tail_sol                                    5
final.output.tail_au                                     0
primary_cleaner.input.sulfate                           20
primary_cleaner.input.depressant                        32
primary_cleaner.input.feed_size                          0
primary_cleaner.input.xanthate                         100
primary_cleaner.output.concentrate_ag                    0
primary_cleaner.output.concentrate_pb                   86
primary_cleaner.output.concentrate_sol                 2

'Статистика:'

final.output.concentrate_ag  final.output.concentrate_pb  \
count                 14148.000000                 14148.000000   
mean                      5.142034                    10.132960   
std                       1.369586                     1.654930   
min                       0.000000                     0.000000   
25%                       4.211620                     9.297355   
50%                       4.994652                    10.297144   
75%                       5.859540                    11.170603   
max                      16.001945                    17.031899   

       final.output.concentrate_sol  final.output.concentrate_au  \
count                  13938.000000                 14149.000000   
mean                       9.202849                    44.003792   
std                        2.790516                     4.905261   
min                        0.000000                     0.000000   
25%                        7.484645                    43.276111   
50%                        8.845462                    44.872436   
75%                       10.487508                    46.166425   
max                       18.124851                    52.756638   

       final.output.recovery  final.output.tail_ag  final.output.tail_pb  \
count           14149.000000          14149.000000          14049.000000   
mean               66.518832              9.607035              2.597298   
std                10.295402              2.319069              0.971843   
min                 0.000000              0.000000              0.000000   
25%                62.545817              7.997429              1.905973   
50%                67.432775              9.480270              2.592022   
75%                72.346428             11.003707              3.241723   
max               100.000000             19.552149              5.639565   

       final.output.tail_sol  final.output.tail_au  \
count           14144.000000          14149.000000   
mean               10.512122              2.918421   
std                 3.003617              0.903712   
min                 0.000000              0.000000   
25%                 8.811324              2.368607   
50%                10.514621              2.851025   
75%                11.933009              3.434764   
max                22.317730              8.197408   

       primary_cleaner.input.sulfate  ...  \
count                   14129.000000  ...   
mean                      133.320659  ...   
std                        39.431659  ...   
min                         0.003112  ...   
25%                       107.006651  ...   
50%                       133.018328  ...   
75%                       159.825396  ...   
max                       250.127834  ...   

       secondary_cleaner.state.floatbank4_a_air  \
count                              14143.000000   
mean                                  19.985454   
std                                    5.657723   
min                                    0.000000   
25%                                   14.990775   
50%                                   20.001789   
75%                                   24.990826   
max                                   30.115735   

       secondary_cleaner.state.floatbank4_a_level  \
count                                14148.000000   
mean                                  -478.696836   
std                                     50.736021   
min                                   -799.709069   
25%                                   -500.628656   
50%                                   -499.681450   
75%                                   -477.472413   
max                                   -245.239184   

       secondary_cleaner.state.floatbank4_b_air  \
count                              14148.000000   
mean                                  15.487065   
std                                    5.255655   
min                                    0.000000   
25%                

'======================================================================================================='

date  primary_cleaner.input.sulfate  \
0  2016-09-01 00:59:59                     210.800909   
1  2016-09-01 01:59:59                     215.392455   
2  2016-09-01 02:59:59                     215.259946   
3  2016-09-01 03:59:59                     215.336236   
4  2016-09-01 04:59:59                     199.099327   
5  2016-09-01 05:59:59                     168.485085   
6  2016-09-01 06:59:59                     144.133440   
7  2016-09-01 07:59:59                     133.513396   
8  2016-09-01 08:59:59                     133.735356   
9  2016-09-01 09:59:59                     126.961069   

   primary_cleaner.input.depressant  primary_cleaner.input.feed_size  \
0                         14.993118                         8.080000   
1                         14.987471                         8.080000   
2                         12.884934                         7.786667   
3                         12.006805                         7.640000   
4                         10.682530                         7.530000   
5                          8.817007                         7.420000   
6                          7.924610                         7.420000   
7                          8.055252                         6.988000   
8                          7.999618                         6.935000   
9                          8.017856                         7.030000   

   primary_cleaner.input.xanthate  primary_cleaner.state.floatbank8_a_air  \
0                        1.005021                             1398.981301   
1                        0.990469                             1398.777912   
2                        0.996043                             1398.493666   
3                        0.863514                             1399.618111   
4                        0.805575                             1401.268123   
5                        0.791191                             1402.826803   
6                        0.788838                             1398.252401   
7                        0.801871                             1401.669677   
8                        0.789329                             1402.358981   
9                        0.805298                             1400.816120   

   primary_cleaner.state.floatbank8_a_level  \
0                               -500.225577   
1                               -500.057435   
2                               -500.868360   
3                               -498.863574   
4                               -500.808305   
5                               -499.299521   
6                               -499.748672   
7                               -501.777839   
8                               -499.981597   
9                               -499.014158   

   primary_cleaner.state.floatbank8_b_air  \
0                             1399.144926   
1                             1398.055362   
2                             1398.860436   
3                             1397.440120   
4                             1398.128818   
5                             1401.511119   
6                             1393.255503   
7                             1400.754446   
8                             1400.985954   
9                             1399.975401   

   primary_cleaner.state.floatbank8_b_level  \
0                               -499.919735   
1                               -499.778182   
2                               -499.764529   
3                               -499.211024   
4                               -499.504543   
5                               -499.205357   
6                               -499.195380   
7                               -502.514024   
8                               -496.802968   
9                               -499.570552   

   primary_cleaner.state.floatbank8_c_air  ...  \
0                             1400.102998  ...   
1                             1396.151033  ...   
2                             1398.075709  ...   
3                

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5290 entries, 0 to 5289
Data columns (total 53 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   date                                        5290 non-null   object 
 1   primary_cleaner.input.sulfate               5286 non-null   float64
 2   primary_cleaner.input.depressant            5285 non-null   float64
 3   primary_cleaner.input.feed_size             5290 non-null   float64
 4   primary_cleaner.input.xanthate              5286 non-null   float64
 5   primary_cleaner.state.floatbank8_a_air      5290 non-null   float64
 6   primary_cleaner.state.floatbank8_a_level    5290 non-null   float64
 7   primary_cleaner.state.floatbank8_b_air      5290 non-null   float64
 8   primary_cleaner.state.floatbank8_b_level    5290 non-null   float64
 9   primary_cleaner.state.floatbank8_c_air      5290 non-null   float64
 10  primary_clea

None

'Кол-во дублей:'

0

'Кол-во НАНов:'

date                                           0
primary_cleaner.input.sulfate                  4
primary_cleaner.input.depressant               5
primary_cleaner.input.feed_size                0
primary_cleaner.input.xanthate                 4
primary_cleaner.state.floatbank8_a_air         0
primary_cleaner.state.floatbank8_a_level       0
primary_cleaner.state.floatbank8_b_air         0
primary_cleaner.state.floatbank8_b_level       0
primary_cleaner.state.floatbank8_c_air         0
primary_cleaner.state.floatbank8_c_level       0
primary_cleaner.state.floatbank8_d_air         0
primary_cleaner.state.floatbank8_d_level       0
rougher.input.feed_ag                          0
rougher.input.feed_pb                          0
rougher.input.feed_rate                        3
rougher.input.feed_size                        1
rougher.input.feed_sol                        21
rougher.input.feed_au                          0
rougher.input.floatbank10_sulfate              5
rougher.input.floatb

'Статистика:'

primary_cleaner.input.sulfate  primary_cleaner.input.depressant  \
count                    5286.000000                       5285.000000   
mean                      174.839652                          8.683596   
std                        43.027080                          3.072050   
min                         2.566156                          0.003839   
25%                       147.121401                          6.489555   
50%                       177.828489                          8.052207   
75%                       208.125438                         10.027764   
max                       265.983123                         40.000000   

       primary_cleaner.input.feed_size  primary_cleaner.input.xanthate  \
count                      5290.000000                     5286.000000   
mean                          7.266339                        1.383803   
std                           0.610219                        0.643474   
min                           5.650000                        0.004984   
25%                           6.890000                        0.907623   
50%                           7.250000                        1.197610   
75%                           7.600000                        1.797819   
max                          15.500000                        4.102454   

       primary_cleaner.state.floatbank8_a_air  \
count                            5.290000e+03   
mean                             1.539494e+03   
std                              1.167979e+02   
min                              5.445860e-32   
25%                              1.498936e+03   
50%                              1.585129e+03   
75%                              1.602077e+03   
max                              2.103104e+03   

       primary_cleaner.state.floatbank8_a_level  \
count                               5290.000000   
mean                                -497.665883   
std                                   19.952431   
min                                 -795.316337   
25%                                 -500.357298   
50%                                 -499.969164   
75%                                 -499.568951   
max                                  -57.195404   

       primary_cleaner.state.floatbank8_b_air  \
count                            5.290000e+03   
mean                             1.545174e+03   
std                              1.222246e+02   
min                              6.647490e-32   
25%                              1.498971e+03   
50%                              1.595622e+03   
75%                              1.602324e+03   
max                              1.813084e+03   

       primary_cleaner.state.floatbank8_b_level  \
count                               5290.000000   
mean                                -500.273098   
std                                   32.968307   
min                                 -799.997015   
25%                                 -500.703002   
50%                                 -500.028514   
75%                                 -499.293257   
max                                 -142.527229   

       primary_cleaner.state.floatbank8_c_air  \
count                            5.290000e+03   
mean                             1.527272e+03   
std                              1.225380e+02   
min                              4.033736e-32   
25%                              1.473230e+03   
50%                              1.549595e+03   
75%                              1.601144e+03   
max                              1.715054e+03   

       primary_cleaner.state.floatbank8_c_level  ...  \
count                               5290.000000  ...   
mean                                -498.330680  ...   
std                                   21.964876  ...   
min                                 -799.960571  ...   
25%                                 -501.018117  ...   
50%                                 -500.017711  ...   
75%                                

'======================================================================================================='

date  final.output.concentrate_ag  \
0  2016-01-15 00:00:00                     6.055403   
1  2016-01-15 01:00:00                     6.029369   
2  2016-01-15 02:00:00                     6.055926   
3  2016-01-15 03:00:00                     6.047977   
4  2016-01-15 04:00:00                     6.148599   
5  2016-01-15 05:00:00                     6.482968   
6  2016-01-15 06:00:00                     6.533849   
7  2016-01-15 07:00:00                     6.130823   
8  2016-01-15 08:00:00                     5.834140   
9  2016-01-15 09:00:00                     5.687063   

   final.output.concentrate_pb  final.output.concentrate_sol  \
0                     9.889648                      5.507324   
1                     9.968944                      5.257781   
2                    10.213995                      5.383759   
3                     9.977019                      4.858634   
4                    10.142511                      4.939416   
5                    10.049416                      5.480257   
6                    10.058141                      4.569100   
7                     9.935481                      4.389813   
8                    10.071156                      4.876389   
9                     9.980404                      5.282514   

   final.output.concentrate_au  final.output.recovery  final.output.tail_ag  \
0                    42.192020              70.541216             10.411962   
1                    42.701629              69.266198             10.462676   
2                    42.657501              68.116445             10.507046   
3                    42.689819              68.347543             10.422762   
4                    42.774141              66.927016             10.360302   
5                    41.633678              69.465816             10.182708   
6                    41.995316              69.300835             10.304598   
7                    42.452727              70.230976             10.443288   
8                    43.404078              69.688595             10.420140   
9                    43.235220              70.279619             10.487013   

   final.output.tail_pb  final.output.tail_sol  final.output.tail_au  ...  \
0              0.895447              16.904297              2.143149  ...   
1              0.927452              16.634514              2.224930  ...   
2              0.953716              16.208849              2.257889  ...   
3              0.883763              16.532835              2.146849  ...   
4              0.792826              16.525686              2.055292  ...   
5              0.664118              16.999638              1.918586  ...   
6              0.807342              16.723575              2.058913  ...   
7              0.949346              16.689959              2.143437  ...   
8              1.065453              17.201948              2.209881  ...   
9              1.159805              17.483979              2.209593  ...   

   secondary_cleaner.state.floatbank4_a_air  \
0                                 14.016835   
1                                 13.992281   
2                                 14.015015   
3                                 14.036510   
4                                 14.027298   
5                                 13.938497   
6                                 14.046819   
7                                 13.974691   
8                                 13.964030   
9                                 13.989632   

   secondary_cleaner.state.floatbank4_a_level  \
0                                 -502.488007   
1                                 -505.503262   
2                                 -502.520901   
3                                 -500.857308   
4                                 -499.838632   
5                                 -500.970168   
6                                 -500.971133   
7                                 -501.819696   
8                             

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19439 entries, 0 to 19438
Data columns (total 87 columns):
 #   Column                                              Non-Null Count  Dtype  
---  ------                                              --------------  -----  
 0   date                                                19439 non-null  object 
 1   final.output.concentrate_ag                         19438 non-null  float64
 2   final.output.concentrate_pb                         19438 non-null  float64
 3   final.output.concentrate_sol                        19228 non-null  float64
 4   final.output.concentrate_au                         19439 non-null  float64
 5   final.output.recovery                               19439 non-null  float64
 6   final.output.tail_ag                                19438 non-null  float64
 7   final.output.tail_pb                                19338 non-null  float64
 8   final.output.tail_sol                               19433 non-null  float64


None

'Кол-во дублей:'

0

'Кол-во НАНов:'

date                                                     0
final.output.concentrate_ag                              1
final.output.concentrate_pb                              1
final.output.concentrate_sol                           211
final.output.concentrate_au                              0
final.output.recovery                                    0
final.output.tail_ag                                     1
final.output.tail_pb                                   101
final.output.tail_sol                                    6
final.output.tail_au                                     0
primary_cleaner.input.sulfate                           24
primary_cleaner.input.depressant                        37
primary_cleaner.input.feed_size                          0
primary_cleaner.input.xanthate                         104
primary_cleaner.output.concentrate_ag                    0
primary_cleaner.output.concentrate_pb                  116
primary_cleaner.output.concentrate_sol                 3

'Статистика:'

final.output.concentrate_ag  final.output.concentrate_pb  \
count                 19438.000000                 19438.000000   
mean                      5.168470                     9.978895   
std                       1.372348                     1.669240   
min                       0.000000                     0.000000   
25%                       4.251240                     9.137262   
50%                       5.066094                    10.102433   
75%                       5.895527                    11.035769   
max                      16.001945                    17.031899   

       final.output.concentrate_sol  final.output.concentrate_au  \
count                  19228.000000                 19439.000000   
mean                       9.501224                    44.076513   
std                        2.787537                     5.129784   
min                        0.000000                     0.000000   
25%                        7.722820                    43.402215   
50%                        9.218961                    45.011244   
75%                       10.947813                    46.275313   
max                       19.615720                    52.756638   

       final.output.recovery  final.output.tail_ag  final.output.tail_pb  \
count           19439.000000          19438.000000          19338.000000   
mean               67.050208              9.688589              2.705795   
std                10.125840              2.328642              0.949077   
min                 0.000000              0.000000              0.000000   
25%                63.299712              8.062878              2.040119   
50%                68.172738              9.743623              2.748730   
75%                72.686642             11.134294              3.333216   
max               100.000000             19.552149              5.804178   

       final.output.tail_sol  final.output.tail_au  \
count           19433.000000          19439.000000   
mean               10.583728              3.042467   
std                 2.868782              0.922808   
min                 0.000000              0.000000   
25%                 8.938947              2.461138   
50%                10.622456              2.984909   
75%                12.104271              3.571351   
max                22.317730              8.245022   

       primary_cleaner.input.sulfate  ...  \
count                   19415.000000  ...   
mean                      144.624774  ...   
std                        44.464071  ...   
min                         0.003112  ...   
25%                       114.107029  ...   
50%                       143.232103  ...   
75%                       175.075656  ...   
max                       265.983123  ...   

       secondary_cleaner.state.floatbank4_a_air  \
count                              19433.000000   
mean                                  18.987674   
std                                    5.411058   
min                                    0.000000   
25%                                   14.975734   
50%                                   18.017481   
75%                                   23.012470   
max                                   30.115735   

       secondary_cleaner.state.floatbank4_a_level  \
count                                19438.000000   
mean                                  -485.894516   
std                                     47.758570   
min                                   -799.798523   
25%                                   -500.704892   
50%                                   -499.837793   
75%                                   -498.249730   
max                                   -245.239184   

       secondary_cleaner.state.floatbank4_b_air  \
count                              19438.000000   
mean                                  15.010366   
std                                    4.890228   
min                                    0.000000   
25%                

'======================================================================================================='

Итого имеем:
* Дублей - нет, что, в принципе, не особо облегчает нашу задачу
* НАНов - куча, с ними надо что-то делать (или не делать) 
* Тестовая выборка не имеет некоторых столбцов, которые имеют полный и тренировочные наборы данных. Для нас это не удивительно, это написано в условиях задачи

## Анализ данных

### Проверка эффективности обогащения

Нужно проверить, что эффективность обогащения (recovery) рассчитана правильно. Будем вычислять её по формуле из ТЗ, на обучающей выборке для признака rougher.output.recovery. Далее найдём MAE между расчётами и значением признака.

In [4]:
# создадим константы для удобства
c = df_train['rougher.output.concentrate_au']
f = df_train['rougher.input.feed_au']
t = df_train['rougher.output.tail_au']

#функиция, которая сделает всё за нас
def recovery_f(c,f,t):
    recovery = (c*(f-t)) / (f*(c-t)) * 100
    return recovery

In [5]:
# посчитаем целевой признак и MAE
recovery = recovery_f(c,f,t)
mae = mean_absolute_error(df_train['rougher.output.recovery'], recovery)
print('Средняя абсолютная ошибка:', mae)

Средняя абсолютная ошибка: 9.73512347450521e-15


Погрешность между расчётным значением и записанным в датафрейме, действительно, погрешность. Получается, нет причин не применять rougher.output.recovery в качестве целевого признака

### Признаки, недоступные в тестовой выборке

In [6]:
# выделим столбцы, которых нет в тестовой выборке, но, которые есть в тренировочных данных
features_not_in_df_test = set(df_train.columns) - set(df_test.columns)
features_not_in_df_test

{'final.output.concentrate_ag',
 'final.output.concentrate_au',
 'final.output.concentrate_pb',
 'final.output.concentrate_sol',
 'final.output.recovery',
 'final.output.tail_ag',
 'final.output.tail_au',
 'final.output.tail_pb',
 'final.output.tail_sol',
 'primary_cleaner.output.concentrate_ag',
 'primary_cleaner.output.concentrate_au',
 'primary_cleaner.output.concentrate_pb',
 'primary_cleaner.output.concentrate_sol',
 'primary_cleaner.output.tail_ag',
 'primary_cleaner.output.tail_au',
 'primary_cleaner.output.tail_pb',
 'primary_cleaner.output.tail_sol',
 'rougher.calculation.au_pb_ratio',
 'rougher.calculation.floatbank10_sulfate_to_au_feed',
 'rougher.calculation.floatbank11_sulfate_to_au_feed',
 'rougher.calculation.sulfate_to_au_concentrate',
 'rougher.output.concentrate_ag',
 'rougher.output.concentrate_au',
 'rougher.output.concentrate_pb',
 'rougher.output.concentrate_sol',
 'rougher.output.recovery',
 'rougher.output.tail_ag',
 'rougher.output.tail_au',
 'rougher.output.ta

Их 34!!!! (но это не плохо и не хорошо). Согласно условию задачи (технологии производства), данные признаки невозможно посчитать заранее и, наверное, их стоит убрать из обучающей выборки. Конечно, кроме целевых - final.output.recovery, rougher.output.recovery. Иначе удачи не видать

### Предобработка данных

Вопрос объёмный, а учитывая количество данных - неподъёмный. Сконцентрируемся на этом:
* посмотрим на аномалии, а далее решим, нужно ли что-то с ними делать
* обработаем пропуски в обучающих данных, пользуясь условием, что близкие по времени данные не сильно отличаются
* добавим целевые признаки для тестовых данных, чтобы было чего потом сравнивать

#### Поиск аномалий

In [ ]:
#построим график распределения значений целевых признаков
fig = go.Figure()
fig.add_trace(go.Histogram(x=df_train['final.output.recovery'], name='"Эффективность обогащения финального концентрата в %"'))
fig.add_trace(go.Histogram(x=df_train['rougher.output.recovery'], name='"Эффективность обогащения чернового концентрата в %"'))
fig.update_layout(barmode='group', 
                  title='Гистограмма распределения извлечения на финальной и флотационной стадиях',
                  xaxis_title='Извлечение, %', yaxis_title='Частота')
fig.show('svg', width=1000, height=500)

Что имеем:
* Нулевые аномалии на обеих стадиях, которые могут означать либо ошибку, либо то, что сырьё пришло "пустое". Поставщик сырья нас обманул
* Есть небольшой всплеск на 100%, но он может означать, что в сырье были самородки (что не редкость в цветной металлургии)

#### Заполнение пропусков и удаление аномалий

In [7]:
# создадим список колонок из тестового набора, но удалим целевые признаки, их мы оставим как есть, нас интересуют пропуски
train_columns = df_train.columns.tolist()
train_columns.remove('final.output.recovery')
train_columns.remove('rougher.output.recovery')

# предварительно отсортируем датасеты по дате
df_train.sort_values(by='date', inplace=True)
df_full.sort_values(by='date', inplace=True)
df_test.sort_values(by='date', inplace=True)

# заполняем только признаки, которые будут участвовать в обучении модели. Делаем всё в полном датафрейме. По условию данные близкие по времени - действительно близки
df_full[train_columns] = df_full[train_columns].ffill(axis = 0)

In [8]:
# индексируем датафреймы по дате, чтобы можно было заменять НАНы далее
df_full = df_full.set_index('date')
df_test = df_test.set_index('date')
df_train = df_train.set_index('date')

In [9]:
# подставляем здоровые данные из полного датафрейма в остальные
df_train.update(df_full)
df_test.update(df_full)

Проверяем

In [10]:
df_train.info()
df_test.info()

<class 'pandas.core.frame.DataFrame'>
Index: 14149 entries, 2016-01-15 00:00:00 to 2018-08-18 10:59:59
Data columns (total 86 columns):
 #   Column                                              Non-Null Count  Dtype  
---  ------                                              --------------  -----  
 0   final.output.concentrate_ag                         14149 non-null  float64
 1   final.output.concentrate_pb                         14149 non-null  float64
 2   final.output.concentrate_sol                        14149 non-null  float64
 3   final.output.concentrate_au                         14149 non-null  float64
 4   final.output.recovery                               14149 non-null  float64
 5   final.output.tail_ag                                14149 non-null  float64
 6   final.output.tail_pb                                14149 non-null  float64
 7   final.output.tail_sol                               14149 non-null  float64
 8   final.output.tail_au                             

На одну проблему меньше

#### Добавление целевых признаков в тестовые данные

In [11]:
# добавим значения из full
df_test['final.output.recovery'] = df_full['final.output.recovery']
df_test['rougher.output.recovery'] = df_full['rougher.output.recovery']

In [12]:
df_test.info()
df_test.describe()

<class 'pandas.core.frame.DataFrame'>
Index: 5290 entries, 2016-09-01 00:59:59 to 2017-12-31 23:59:59
Data columns (total 54 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   primary_cleaner.input.sulfate               5290 non-null   float64
 1   primary_cleaner.input.depressant            5290 non-null   float64
 2   primary_cleaner.input.feed_size             5290 non-null   float64
 3   primary_cleaner.input.xanthate              5290 non-null   float64
 4   primary_cleaner.state.floatbank8_a_air      5290 non-null   float64
 5   primary_cleaner.state.floatbank8_a_level    5290 non-null   float64
 6   primary_cleaner.state.floatbank8_b_air      5290 non-null   float64
 7   primary_cleaner.state.floatbank8_b_level    5290 non-null   float64
 8   primary_cleaner.state.floatbank8_c_air      5290 non-null   float64
 9   primary_cleaner.state.floatbank8_c_level    5290 non-null

primary_cleaner.input.sulfate  primary_cleaner.input.depressant  \
count                    5290.000000                       5290.000000   
mean                      174.842478                          8.684695   
std                        43.023252                          3.073753   
min                         2.566156                          0.003839   
25%                       147.121401                          6.489556   
50%                       177.828489                          8.052131   
75%                       208.125438                         10.028043   
max                       265.983123                         40.000000   

       primary_cleaner.input.feed_size  primary_cleaner.input.xanthate  \
count                      5290.000000                     5290.000000   
mean                          7.266339                        1.383430   
std                           0.610219                        0.643548   
min                           5.650000                        0.004984   
25%                           6.890000                        0.907390   
50%                           7.250000                        1.197586   
75%                           7.600000                        1.797819   
max                          15.500000                        4.102454   

       primary_cleaner.state.floatbank8_a_air  \
count                            5.290000e+03   
mean                             1.539494e+03   
std                              1.167979e+02   
min                              5.445860e-32   
25%                              1.498936e+03   
50%                              1.585129e+03   
75%                              1.602077e+03   
max                              2.103104e+03   

       primary_cleaner.state.floatbank8_a_level  \
count                               5290.000000   
mean                                -497.665883   
std                                   19.952431   
min                                 -795.316337   
25%                                 -500.357298   
50%                                 -499.969164   
75%                                 -499.568951   
max                                  -57.195404   

       primary_cleaner.state.floatbank8_b_air  \
count                            5.290000e+03   
mean                             1.545174e+03   
std                              1.222246e+02   
min                              6.647490e-32   
25%                              1.498971e+03   
50%                              1.595622e+03   
75%                              1.602324e+03   
max                              1.813084e+03   

       primary_cleaner.state.floatbank8_b_level  \
count                               5290.000000   
mean                                -500.273098   
std                                   32.968307   
min                                 -799.997015   
25%                                 -500.703002   
50%                                 -500.028514   
75%                                 -499.293257   
max                                 -142.527229   

       primary_cleaner.state.floatbank8_c_air  \
count                            5.290000e+03   
mean                             1.527272e+03   
std                              1.225380e+02   
min                              4.033736e-32   
25%                              1.473230e+03   
50%                              1.549595e+03   
75%                              1.601144e+03   
max                              1.715054e+03   

       primary_cleaner.state.floatbank8_c_level  ...  \
count                               5290.000000  ...   
mean                                -498.330680  ...   
std                                   21.964876  ...   
min                                 -799.960571  ...   
25%                                 -501.018117  ...   
50%                                 -500.017711  ...   
75%                                

Есть 0-вые значения по целевым признакам, использовать их мы конечно же не будем. Будем дропать в следующем блоке

In [ ]:
# убираем аномалии
#df_full = df_full.loc[(df_full['final.output.recovery'] > 0) & (df_full['rougher.output.recovery'] > 0)]
#df_train = df_train.loc[(df_train['final.output.recovery'] > 0) & (df_train['rougher.output.recovery'] > 0)]
#df_test = df_test.loc[(df_test['final.output.recovery'] > 0) & (df_test['rougher.output.recovery'] > 0)]
#display(df_full.describe())
#display(df_train.describe())
#display(df_test.describe())

Можно двигаться дальше

### Анализ данных

#### Концентрация металлов (Au, Ag, Pb) на различных этапах очистки

In [ ]:
# создадим словарь с интересующими нас объектами
metals = {'au':['rougher.input.feed_au', 'rougher.output.concentrate_au', 'primary_cleaner.output.concentrate_au', 'final.output.concentrate_au' ], 'ag':['rougher.input.feed_ag', 'rougher.output.concentrate_ag', 'primary_cleaner.output.concentrate_ag', 'final.output.concentrate_ag' ], 'pb':['rougher.input.feed_pb', 'rougher.output.concentrate_pb', 'primary_cleaner.output.concentrate_pb', 'final.output.concentrate_pb' ], 'sol':['rougher.input.feed_sol', 'rougher.output.concentrate_sol', 'primary_cleaner.output.concentrate_sol', 'final.output.concentrate_sol']}
# все графики построим в цикле
for metal in metals:
    fig = px.box(df_train[metals[metal]], title=f'Концентрация {metal} на разных стадиях обогащения')
    fig.show()

Au - по продвижению к конечному продукту, концентрация возрастает. Собственно, ради этого всё и затевалось на этом заводе

Ag - по отношению к золоту является примесью, концентрация в конечном продукте падает,вероятно попадает в "хвосты" и извлекается далее в рамках другого технологического процесса.

Pb - на продвижение по стадиям особо не реагирует. Либо особенности технологического процесса и химии, либо добавляется искуственно, что опять же является особенностью

Sol - остальные примеси также в очень большом количестве отлетают после первого этапа, далее эффективность процесса снижается

#### Распределение размера гранул сырья на обучающей и тестовых выборках

In [ ]:
# сгруппируем данные для plotly
hist_data = [df_test['rougher.input.feed_size'], df_train['rougher.input.feed_size']]
group_labels = ['Распределние размера на тестовой выборке', 'Распределние размера гранул на обучающей выборке']

# построим график
fig = ff.create_distplot(hist_data, group_labels, bin_size=5)
fig.update_layout(title_text='Распределение размера гранул сырья')
fig.show()

На стадии флотации, размеры гранул сырья хоть и сильно разнятся, но, в среднем, имеют похожую форму распределния по частоте и стараются быть 'нормальными'. Каких-то супер проблем это доставить не должно 

#### Cуммарая концентрация веществ на разных стадиях

In [14]:
sum_metals = {'rougher.input':['rougher.input.feed_ag', 'rougher.input.feed_pb', 'rougher.input.feed_sol', 'rougher.input.feed_au'],
    'rougher.output':[x for x in df_full.columns if x.startswith('rougher.output.concentrate')],
               'primary.cleaner':[x for x in df_full.columns if x.startswith('primary_cleaner.output.concentrate')],
               'final':[x for x in df_full.columns if x.startswith('final.output.concentrate')]
                  }

for metal in sum_metals:
    hist_data = [np.sum(df_full[sum_metals[metal]], axis = 1)]
    group_labels = ['Концентрация']
    fig = ff.create_distplot(hist_data, group_labels)
    fig.update_layout(title_text=f'Концентрация веществ на стадии {metal}')
    fig.show()

Имеем следующее:
1. С переходом на более высокую стадию, гистограмма распределения суммарного количества веществ сжимается, что говорит нам о том, что всё лишнее постепенно удаляется из раствора
2. На стадии primary_cleaner есть небольшой всплеск значений около нуля. Вряд ли такое может быть, так что, скорее всего это аномалии. Попробуем удалить, если что-то пойдёт не так, то всегда можно вернуть на место
3. Также помним, что есть 0 значения по целевым признакам, от них также будем избавляться

#### Аномалии

In [15]:
# удаляем
df_full = df_full[np.sum(df_full[sum_metals['primary.cleaner']], axis = 1) > 0]
# заменяем данные в остальных датасетах
df_train = df_train.merge(df_full, how='inner')
df_test = df_test.merge(df_full[df_test.columns], how='inner')
# профит

In [16]:
# борьба с нулевыми значениями в целевых признаках
df_full = df_full.loc[(df_full['final.output.recovery'] > 0) & (df_full['rougher.output.recovery'] > 0)]
df_train = df_train.loc[(df_train['final.output.recovery'] > 0) & (df_train['rougher.output.recovery'] > 0)]
df_test = df_test.loc[(df_test['final.output.recovery'] > 0) & (df_test['rougher.output.recovery'] > 0)]
display(df_full.describe())
display(df_train.describe())
display(df_test.describe())

final.output.concentrate_ag  final.output.concentrate_pb  \
count                 18848.000000                 18848.000000   
mean                      5.198960                    10.030318   
std                       1.326194                     1.510454   
min                       0.010000                     0.010000   
25%                       4.271088                     9.148408   
50%                       5.079826                    10.110375   
75%                       5.898205                    11.049269   
max                      16.001945                    17.031899   

       final.output.concentrate_sol  final.output.concentrate_au  \
count                  18848.000000                 18848.000000   
mean                       9.489024                    44.292759   
std                        2.806800                     4.014120   
min                        0.010000                     0.010000   
25%                        7.680258                    43.424311   
50%                        9.206063                    45.010693   
75%                       10.961220                    46.262709   
max                       19.615720                    52.756638   

       final.output.recovery  final.output.tail_ag  final.output.tail_pb  \
count           18848.000000          18848.000000          18848.000000   
mean               67.437596              9.700515              2.692079   
std                 8.866445              2.307169              0.953020   
min                 0.168611              0.000000              0.000000   
25%                63.513256              8.058638              2.027906   
50%                68.282446              9.757114              2.732702   
75%                72.744702             11.140695              3.318849   
max               100.000000             19.552149              5.804178   

       final.output.tail_sol  final.output.tail_au  \
count           18848.000000          18848.000000   
mean               10.598722              3.032764   
std                 2.850230              0.913306   
min                 0.000000              0.000000   
25%                 8.933251              2.456036   
50%                10.631250              2.973366   
75%                12.096591              3.554258   
max                22.317730              8.245022   

       primary_cleaner.input.sulfate  ...  \
count                   18848.000000  ...   
mean                      145.188881  ...   
std                        44.106244  ...   
min                         0.051279  ...   
25%                       114.718648  ...   
50%                       143.618265  ...   
75%                       175.307881  ...   
max                       265.983123  ...   

       secondary_cleaner.state.floatbank4_a_air  \
count                              1.884800e+04   
mean                               1.904006e+01   
std                                5.424415e+00   
min                                1.079872e-16   
25%                                1.498029e+01   
50%                                1.801906e+01   
75%                                2.301726e+01   
max                                3.011574e+01   

       secondary_cleaner.state.floatbank4_a_level  \
count                                18848.000000   
mean                                  -485.728804   
std                                     47.647695   
min                                   -799.798523   
25%                                   -500.703483   
50%                                   -499.839909   
75%                                   -498.249898   
max                                   -245.239184   

       secondary_cleaner.state.floatbank4_b_air  \
count                              1.884800e+04   
mean                               1.505687e+01   
std                                4.881833e+00   
min                                2.489718e-17   
25%                

final.output.concentrate_ag  final.output.concentrate_pb  \
count                 13725.000000                 13725.000000   
mean                      5.164996                    10.170553   
std                       1.332867                     1.522093   
min                       0.010000                     0.010000   
25%                       4.223187                     9.298153   
50%                       5.004025                    10.303898   
75%                       5.859495                    11.181332   
max                      16.001945                    17.031899   

       final.output.concentrate_sol  final.output.concentrate_au  \
count                  13725.000000                 13725.000000   
mean                       9.154032                    44.188519   
std                        2.844241                     3.943532   
min                        0.010000                     0.010000   
25%                        7.428656                    43.290494   
50%                        8.815563                    44.871192   
75%                       10.471206                    46.149523   
max                       18.124851                    52.756638   

       final.output.recovery  final.output.tail_ag  final.output.tail_pb  \
count           13725.000000          13725.000000          13725.000000   
mean               66.808927              9.622242              2.578954   
std                 9.286365              2.294249              0.977892   
min                 0.168611              0.000000              0.000000   
25%                62.738502              7.996374              1.889206   
50%                67.468279              9.500774              2.566671   
75%                72.373813             11.013138              3.221016   
max               100.000000             19.552149              5.639565   

       final.output.tail_sol  final.output.tail_au  \
count           13725.000000          13725.000000   
mean               10.523165              2.911019   
std                 2.988293              0.897626   
min                 0.000000              0.000000   
25%                 8.805357              2.364556   
50%                10.522160              2.840130   
75%                11.919450              3.420005   
max                22.317730              8.197408   

       primary_cleaner.input.sulfate  ...  \
count                   13725.000000  ...   
mean                      133.865293  ...   
std                        39.090577  ...   
min                         0.051279  ...   
25%                       107.775541  ...   
50%                       133.469157  ...   
75%                       159.984727  ...   
max                       250.127834  ...   

       secondary_cleaner.state.floatbank4_a_air  \
count                              13725.000000   
mean                                  20.035697   
std                                    5.691333   
min                                    0.000073   
25%                                   14.993173   
50%                                   20.005098   
75%                                   24.992343   
max                                   30.115735   

       secondary_cleaner.state.floatbank4_a_level  \
count                                13725.000000   
mean                                  -478.486348   
std                                     50.621769   
min                                   -799.706757   
25%                                   -500.626380   
50%                                   -499.680971   
75%                                   -467.534188   
max                                   -245.239184   

       secondary_cleaner.state.floatbank4_b_air  \
count                              13725.000000   
mean                                  15.524251   
std                                    5.267992   
min                                    0.000068   
25%                

primary_cleaner.input.sulfate  primary_cleaner.input.depressant  \
count                    5123.000000                       5123.000000   
mean                      175.525841                          8.666259   
std                        42.423213                          3.065337   
min                         2.926720                          0.057383   
25%                       147.859719                          6.486437   
50%                       178.676656                          8.039751   
75%                       208.310366                         10.017786   
max                       265.983123                         40.000000   

       primary_cleaner.input.feed_size  primary_cleaner.input.xanthate  \
count                      5123.000000                     5123.000000   
mean                          7.258789                        1.387709   
std                           0.605351                        0.632937   
min                           5.650000                        0.004984   
25%                           6.890000                        0.918104   
50%                           7.250000                        1.199125   
75%                           7.590000                        1.797875   
max                          15.500000                        4.102454   

       primary_cleaner.state.floatbank8_a_air  \
count                            5.123000e+03   
mean                             1.542630e+03   
std                              1.121826e+02   
min                              5.445860e-32   
25%                              1.499299e+03   
50%                              1.591649e+03   
75%                              1.602156e+03   
max                              1.810767e+03   

       primary_cleaner.state.floatbank8_a_level  \
count                               5123.000000   
mean                                -497.664103   
std                                   19.530878   
min                                 -783.448551   
25%                                 -500.347936   
50%                                 -499.969714   
75%                                 -499.575528   
max                                  -57.195404   

       primary_cleaner.state.floatbank8_b_air  \
count                            5.123000e+03   
mean                             1.548953e+03   
std                              1.186775e+02   
min                              6.647490e-32   
25%                              1.499303e+03   
50%                              1.596481e+03   
75%                              1.602468e+03   
max                              1.813084e+03   

       primary_cleaner.state.floatbank8_b_level  \
count                               5123.000000   
mean                                -499.619511   
std                                   29.951302   
min                                 -799.950103   
25%                                 -500.692253   
50%                                 -500.026610   
75%                                 -499.288840   
max                                 -142.527229   

       primary_cleaner.state.floatbank8_c_air  \
count                            5.123000e+03   
mean                             1.531389e+03   
std                              1.189574e+02   
min                              4.033736e-32   
25%                              1.484034e+03   
50%                              1.550184e+03   
75%                              1.601323e+03   
max                              1.715054e+03   

       primary_cleaner.state.floatbank8_c_level  ...  \
count                               5123.000000  ...   
mean                                -498.295525  ...   
std                                   21.604527  ...   
min                                 -799.960571  ...   
25%                                 -501.011189  ...   
50%                                 -500.018023  ...   
75%                                

#### Выводы по этапу

1.	Концентрация Au растёт с переходом на следующие этапы очистки. Концентрация примесей падает - частный случай корреляции
2.	Распределение размеров гранул сырья на обучающей и тестовых выборках имеют схожую форму и вполне сопоставимый масштаб.
3.	С переходом на более высокую стадию, гистограмма распределения суммарного количества веществ сжимается, что говорит нам о том, что всё лишнее постепенно удаляется из раствора
4.	На стадии primary_cleaner имелся небольшой всплеск значений около нуля. Вряд ли такое может быть, так что, скорее всего это аномалии. Их мы удалили и произвели замену в остальных датасетах

## Модель

### Функция для вычисления sMAPE

In [17]:
def smape(target, predicted):
    return ((1/len(target)) * np.sum(2 * np.abs(target - predicted) / (np.abs(target) + np.abs(predicted)))) * 100

### Обучение моделей

Подготовим признаки для обучения модели

In [18]:
list_test_columns = df_test.columns
# подготовим  обучающие признаки для этапа флотации
features_rougher = []
for column in list_test_columns:
    if "rougher" in column:
        features_rougher.append(column)

features_rougher_train = df_train[features_rougher]
features_rougher_train = features_rougher_train.drop('rougher.output.recovery', axis=1)
target_rougher_train = df_train['rougher.output.recovery']
display(features_rougher_train.head())
display(target_rougher_train.head())

rougher.input.feed_ag  rougher.input.feed_pb  rougher.input.feed_rate  \
0               6.100378               2.284912               523.546326   
1               6.161113               2.266033               525.290581   
2               6.116455               2.159622               530.026610   
3               6.043309               2.037807               542.590390   
4               6.060915               1.786875               540.531893   

   rougher.input.feed_size  rougher.input.feed_sol  rougher.input.feed_au  \
0                55.486599               36.808594               6.486150   
1                57.278666               35.753385               6.478583   
2                57.510649               35.971630               6.362222   
3                57.792734               36.862241               6.118189   
4                56.047189               34.347666               5.663707   

   rougher.input.floatbank10_sulfate  rougher.input.floatbank10_xanthate  \
0                          11.986616                            6.007990   
1                          11.971193                            6.005766   
2                          11.920603                            6.197377   
3                          11.630094                            6.203177   
4                          10.957755                            6.198826   

   rougher.input.floatbank11_sulfate  rougher.input.floatbank11_xanthate  ...  \
0                          11.836743                            6.005818  ...   
1                          11.996163                            6.012594  ...   
2                          11.920305                            6.204633  ...   
3                          11.692450                            6.196578  ...   
4                          10.960521                            6.194897  ...   

   rougher.state.floatbank10_b_air  rougher.state.floatbank10_b_level  \
0                      1603.011353                        -434.715027   
1                      1600.754587                        -449.953435   
2                      1599.337330                        -450.008530   
3                      1600.059442                        -450.619948   
4                      1600.208824                        -449.599614   

   rougher.state.floatbank10_c_air  rougher.state.floatbank10_c_level  \
0                      1602.375000                        -442.204468   
1                      1600.479580                        -449.830646   
2                      1599.672797                        -449.954491   
3                      1600.012842                        -449.910497   
4                      1600.357732                        -450.034364   

   rougher.state.floatbank10_d_air  rougher.state.floatbank10_d_level  \
0                      1598.937256                        -451.294128   
1                      1600.527589                        -449.953649   
2                      1599.849325                        -449.954185   
3                      1597.725177                        -450.130127   
4                      1599.759049                        -449.909799   

   rougher.state.floatbank10_e_air  rougher.state.floatbank10_e_level  \
0                      1404.472046                        -455.462982   
1                      1399.227084                        -450.869848   
2                      1399.180945                        -449.937588   
3                      1400.943157                        -450.030142   
4                      1401.560902                        -448.877187   

   rougher.state.floatbank10_f_air  rougher.state.floatbank10_f_level  
0                      1416.354980                        -451.939636  
1                      1399.719514                        -450.119001  
2                      1400.316682                        -450.527147  
3                      1400.234743                        -449.790835  
4                      

0    87.107763
1    86.843261
2    86.842308
3    87.226430
4    86.688794
Name: rougher.output.recovery, dtype: float64

In [19]:
# подготовим  обучающие признаки для финального этапа
features_final_train = df_train[df_test.columns]
features_final_train = features_final_train.drop('final.output.recovery', axis=1)
target_final_train = df_train['final.output.recovery']

# подготовим тестовые выборки
features_rougher_test = df_test[features_rougher]
features_rougher_test = features_rougher_test.drop('rougher.output.recovery', axis=1)
target_rougher_test = df_test['rougher.output.recovery']
features_final_test = df_test.drop('final.output.recovery', axis=1)
target_final_test = df_test['final.output.recovery']

In [21]:
display(features_rougher_train.info())
#display(target_final_train.info())
display(features_final_test.info())
#display(target_final_test.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 13725 entries, 0 to 14074
Data columns (total 22 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   rougher.input.feed_ag               13725 non-null  float64
 1   rougher.input.feed_pb               13725 non-null  float64
 2   rougher.input.feed_rate             13725 non-null  float64
 3   rougher.input.feed_size             13725 non-null  float64
 4   rougher.input.feed_sol              13725 non-null  float64
 5   rougher.input.feed_au               13725 non-null  float64
 6   rougher.input.floatbank10_sulfate   13725 non-null  float64
 7   rougher.input.floatbank10_xanthate  13725 non-null  float64
 8   rougher.input.floatbank11_sulfate   13725 non-null  float64
 9   rougher.input.floatbank11_xanthate  13725 non-null  float64
 10  rougher.state.floatbank10_a_air     13725 non-null  float64
 11  rougher.state.floatbank10_a_level   13725

None

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5123 entries, 0 to 5255
Data columns (total 53 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   primary_cleaner.input.sulfate               5123 non-null   float64
 1   primary_cleaner.input.depressant            5123 non-null   float64
 2   primary_cleaner.input.feed_size             5123 non-null   float64
 3   primary_cleaner.input.xanthate              5123 non-null   float64
 4   primary_cleaner.state.floatbank8_a_air      5123 non-null   float64
 5   primary_cleaner.state.floatbank8_a_level    5123 non-null   float64
 6   primary_cleaner.state.floatbank8_b_air      5123 non-null   float64
 7   primary_cleaner.state.floatbank8_b_level    5123 non-null   float64
 8   primary_cleaner.state.floatbank8_c_air      5123 non-null   float64
 9   primary_cleaner.state.floatbank8_c_level    5123 non-null   float64
 10  primary_clea

None

In [ ]:
# для кросс-валидации тоже создадим функцию
def cross_val_func (model, features, target):
    scores = cross_val_score(model, features, 
                                   target, scoring=make_scorer(smape, greater_is_better=True), cv=5,n_jobs=-1)
    smape_result = np.mean(scores)
    return smape_result

In [ ]:
# для подсчёта финальной метрики sMAPE
def smape_full(model):
    smape_rougher = cross_val_func(model, features_rougher_train, target_rougher_train)
    smape_final = cross_val_func(model, features_final_train, target_final_train)
    smape_rf = 0.25 * smape_rougher + 0.75 * smape_final
    print('sMAPE c кросс-валидацией на этапе флотации:', smape_rougher)
    print('sMAPE c кросс-валидацией на финальном этапе:', smape_final)
    print('Финальная sMAPE c кросс-валидацией составила', smape_rf)

При выборе модели предлагаю использовать 4 классических алгоритма: линейная регрессия, случайный лес, решающее дерево и, для сравнения, dummy регрессия. Далее, для лучшей модели попробуем подобрать оптимальные гиперпараметры и ещё улучшить результат метрики

In [ ]:
model = LinearRegression()
smape_full(model)

In [ ]:
model = RandomForestRegressor()
smape_full(model)

In [ ]:
model = DecisionTreeRegressor()
smape_full(model)

In [ ]:
model = DummyRegressor()
smape_full(model)

Лучше всех на данном этапе предсказывает решающее дерево, попробуем улучшить его с помощью подбора гиперпараметров

### Подбор гиперпараметров

In [ ]:
best_smape = 10
best_depth = 0
for depth in range (1, 15):
    model = DecisionTreeRegressor (max_depth=depth)
    smape_rougher = cross_val_func (model, features_rougher_train, target_rougher_train)
    smape_final = cross_val_func (model, features_final_train, target_final_train)
    smape_last = 0.25 * smape_rougher + 0.75 * smape_final
    if smape_last < best_smape:
        best_smape = smape_last
        best_depth = depth
print('Наилучшая модель с sMAPE, равным {}, получена для решающего дерева с глубиной {}'.format(best_smape, best_depth))

In [ ]:
best_smape = 10
best_est = 0
best_depth = 0
for est in range (10, 51, 10):
    for depth in range (1, 5):
        model = RandomForestRegressor(n_estimators=est, max_depth=depth)
        smape_rougher = cross_val_func (model, features_rougher_train, target_rougher_train)
        smape_final = cross_val_func (model, features_final_train, target_final_train)
        smape_last = 0.25 * smape_rougher + 0.75 * smape_final
        if smape_last < best_smape:
            best_smape = smape_last
            best_est = est
            best_depth = depth
print('Наилучшая модель с sMAPE, равным {}, получена для случайного леса \
с количеством деревьев {} и глубиной {}'.format(best_smape, best_est, best_depth))

Подбор гиперпараметров в разумных диапазонах значений параметров не даёт увеличение метрики. Для проверки на тестовых данных предлагаю использовать данные модели с дефолтными гиперпараметрами. 

### Проверка на тестовых данных

In [22]:
# сделаем аналогичную функцию для теста
def smape_full_test(model):
    model_rougher = model
    model_rougher.fit(features_rougher_train, target_rougher_train)
    predicted_rougher = model_rougher.predict(features_rougher_test)
    

    model_final = model
    model_final.fit(features_final_train, target_final_train)
    predicted_final = model_final.predict(features_final_test)
    
    smape_rougher = smape(target_rougher_test, predicted_rougher)
    smape_final = smape(target_final_test, predicted_final)
    smape_test = 0.25 * smape_rougher + 0.75 * smape_final
    print('sMAPE тестовых данных на этапе флотации:', smape_rougher)
    print('sMAPE тестовых данных на финальном этапе', smape_final)
    print('Конечная sMAPE', smape_test)

In [23]:
model = DecisionTreeRegressor()
smape_full_test(model)

sMAPE тестовых данных на этапе флотации: 7.544068608729063
sMAPE тестовых данных на финальном этапе 10.50681759143363
Конечная sMAPE 9.766130345757489


In [25]:
model = RandomForestRegressor(n_estimators=31, max_depth=4)
smape_full_test(model)

sMAPE тестовых данных на этапе флотации: 4.799494410715379
sMAPE тестовых данных на финальном этапе 6.389799269729528
Конечная sMAPE 5.99222305497599


In [24]:
model = DummyRegressor()
smape_full_test(model)

sMAPE тестовых данных на этапе флотации: 5.364780952068203
sMAPE тестовых данных на финальном этапе 8.58252552273144
Конечная sMAPE 7.778089380065631


## Общий вывод

В процессе работы над проектом проделано следующее:
* Проведён исследовательский анализ данных
* Проведена предобработка данных
* Проведён анализ данных по целевым направлениям

Также обучена модель решающего дерева для пресказания целевых признаков.
В целом, можно заключить, что модель решающего дерева справляется со своей задачей. Финальная метрика sMAPE на тестовых данных составила 9.77 против 7.77 константной модели